### ISSUE TITLE
Feature request - Folder Depth

### ISSUE THREAD

**Issue by starsoccer** [eneam/mboxviewer#14]
It would be nice if mboxviewer could be used to open a top level folder and automatically open sub folders for eml files.

In the below example I should be able to just open the top level `Emails` folder and then have all eml files below even within multiple folders come up.

Example folder structure:
```
- A
- B
- Emails
 - Folder1
  - email1.eml
  - email2.eml
  - SubFolder1
   - email11.eml
 - Folder2
  - email5.eml
  - email6.eml
 - Folder3
- C
```

**Comment by zigm** [comment]
Thanks for suggestion. Expansion could happen upon one time explicit user request via menu option.
Current database maintained by the mbox viewer is fairly simple and needs to be enhanced to accommodate request you raised and other similar requests such as showing virtual folders such as Sent, Receive, etc per mbox mail.
I will consider your request along with other similar requests or sooner if possible after I analyze what would have to done.

Thanks,
Zbigniew

**Comment by starsoccer** [comment]
okay cool thanks appreciate you being so responsive

**Comment by mattesh** [comment]
Hi Zbigniew,
just want to put thumbs up for this feature.
Did an export and have several sub folders where only second++ level contains the mbox files.
It's pain to load all this files in the folder list one after the other.
e.g.
mails
+--- projects
        +--- project_a
         |      +--- offer
         |      |       offer.mbox
         |      +--- contract
         |      |       contract.mbox
         |      +--- reporting
         |      |       reporting.mbox
         |      +--- features
         |      |       features.mbox
        +--- project_b
         |      +--- offer
         |      |       offer.mbox
         |      +--- contract
         |      |       contract.mbox
         |      +--- reporting
         |      |       reporting.mbox
         |      +--- features
         |      |       features.mbox

Can I avoid this manual effort somehow?

thanks
Mattes

**Comment by zigm** [comment]
I agree with your view on this missing feature. It is  probably  not a large effort to iterate root folder and create flat view of the root folders. I guess that is what you see now after manual steps, i.e. multiple "features" folders at the same level. Such a view is not optimal, hierarchy would be more user friendly so "features" folder would appear under the project_a and project_b folders.

Creating sub-folders may not be too difficult either but there will be impact on many other features supported by MBox Viewer. I will start prototyping in order to evaluate the scope of required changes. Less changes, the lower the risk of breaking something and less time needed to do the regression test.

I will start prototyping and provide updates on the progress.

Q: you exported mail data from what mail service using what tool or mail client?

thanks
Zbigniew

**Comment by DHSch** [comment]
Hi. I would also love this feature of sub-folders. I exported a Lotus Notes mail database using [NSF2X](https://github.com/adb014/nsf2x). Now I have 340 mbox files in 82 nested sub-folders. 
mboxviewer works really well with these mbox files! But it is a  lot of effort to go back to the "select folder" pop-up window every time to change the folder.

thanks,
Dirk

**Comment by zigm** [comment]
Hi Dirk,

Some time ago, I implemented several enhancements to support exported Thunderbird mail archives but didn't make official package. New menu option "File->Select root folder" will traverse the user selected root folder and all sub-folders looking for valid mail archives.  It will automatically create folders under the Mail Tree for root and all sub-folders with mail archives present. However, all sub-folders are installed under the same level, no hierarchy of folders.

In your case, MBox Viewer would automatically create 82 folders under the same level. Not ideal in many cases since related mails grouping/hierarchy  is lost. I am looking now what code changes would be required to support folder hierarchy. 

Are you familiar with Gmail and Labels?  Are exported Lotus Notes sub-folders similar to Gmail Label?

Also, can you provide additional comment on  your statement below:

**But it is a lot of effort to go back to the "select folder" pop-up window every time to change the folder.**

MBox Viewer allows to install multiply folders under the Mail Tree. You should be able to install all 82 sub-folders and not have to go back to the "select folder" pop-up window every time to change the folder.  However, folder hierarchy will be lost. Also, you still need to "select folder" option once per each sub-folder. The 'Select root folder" option would work better somewhat.

Thanks for your comment and suggestions.

Thank you,
Zbigniew

**Comment by DHSch** [comment]
Hi Zbigniew,
yes, you are right. And I did not describe it well. I only need to go back to the "select folder" window at the beginning, to select each sub-folder once. But the more important problem is the lost folder hierarchy, as you said.

I am familiar with how labels are used in the GMail UI, but not how they are represented in exported mbox files.
What I can tell you: The exported Lotus Notes sub-folder hierarchy is represented using regular windows folders. 
Example: I have three mail folders in Lotus Notes called top/sub1/sub2 , and each one also contains mails. Then the export will have the following folders and files in the file system:
top.mbox
top\
top\sub1.mbox
top\sub1\
top\sub1\sub2.mbox

If there were no mails in a folder (but only sub-folders), then the mbox file would be missing, but the sub-folder would still be there. 

I think this is very similar to Thunderbird. Only Thunderbird does not have ".mbox" extensions for files, but it has ".sbd" extensions for subfolders. So the same filesystem layout in Thunderbird will be like this:

top
top.sbd\
top.sbd\sub1
top.sbd\sub1.sbd\
top.sbd\sub1.sbd\sub2

cheers,
Dirk

**Comment by starsoccer** [comment]
+1 on keeping the hierarchy if possible. Honestly something as simple as just prefixing emails with a -> might be sufficient. Right now for me atleast the biggest pain point is just as previously mentioned having to open each folder rather then being able to select a root folder.

**Comment by zigm** [comment]
To support Thunderbird mail files, I enhanced MBox Viewer to examine and validate all files in the folder regardless of the extension. Solution is a bit more costly if there are large number of non-mbox files present in the folder(s).

I will provide update on the progress to support true or simplified folder hierarchy.

**Comment by zigm** [comment]
I released v1.0.3.28 to support exports from Thunderbird as you requested I believe.

The new MBox Viewer will merge all mbox files under root folder and sub-folders and automatically recreate folder structure as Gmail style labels.

Select "File-->Select root folder for merging.." option to automatically import all files based on the root folder.

Appending extension is no longer required in this release. If you decide to try the latest version, let me know if it worked for you and suggest enhancements if needed.

Note that this version of MBox Viewer will delete all files generated by currently installed MBox Viewer except for index files. During startup MBox Viewer will ask you whether it is ok to proceed. You may select NO if you wish to examine current files created by "Print To" or exports and save some if needed.

MBox Viewer will also ask you to configure data folder for all files created by new MBox Viewer.

README file lists all new features and changes implemented in v1.0.3.28.

**Comment by zigm** [comment]
I implemented "Select folder as root folder" option first. Later I realized that merging files from the same account is probably a better option since you it allows searching all emails more easily instead of searching one mbox file at a time. "Select folder as root folder" option could obviously be enhanced to allow searching emails in all folders but it doesn't work like this currently.

Thanks for the feedback. If you decide to try the merging option, let me know if it worked for you.
